server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java

```diff
@@ -258,4 +257,27 @@ private static boolean hasNonFrozen(DiscoveryNode node) {
         return node.getRoles().stream().anyMatch(r -> r.canContainData() && r != DiscoveryNodeRole.DATA_FROZEN_NODE_ROLE);
     }
 
+    static String errorMessageFrom(Result result) {
+        return "this action would add ["
+            + result.totalShardsToAdd
+            + "] shards, but this cluster currently has ["
+            + result.currentUsedShardsByGroup.get()
+            + "]/["
+            + result.maxShardsInCluster
+            + "] maximum "
+            + result.group
+            + " shards open";
+    }
+
+    /**
+     * A Result object containing enough information to be used by external callers about the state of the cluster from the shard limits
+     * perspective.
+     */
+    public record Result(
+        boolean canAddShards,
+        Optional<Long> currentUsedShardsByGroup,
+        int totalShardsToAdd,
+        int maxShardsInCluster,
```

Comment:
This is also per group, right? Is it worth clarifying that (since `currentUsedShardsByGroup` does)?

Reply:
right! removed the suffix, since they `Result` is already namespaced by group